File: leetcode/Subsets_2.cpp
class Solution {
public:
    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        if(S.empty()) return vector<vector<int>>();
        
        vector<vector<int>> result;
        
        sort(S.begin(), S.end());
        vector<int> t;
        subset_helper(result, t, S);
        return result;
    }
private:
    void subset_helper(vector<vector<int>> &result, vector<int> &sofar, vector<int> &rest){
        if(rest.empty()){
            result.push_back(sofar);
            return;
        }
        
        int x = rest[0];
        sofar.push_back(x);
        rest.erase(rest.begin());
        subset_helper(result, sofar, rest);
        
        int cnt = 1;
        while(!rest.empty() && x == rest[0]){
            rest.erase(rest.begin());
            cnt++;
        }
        sofar.pop_back();
        subset_helper(result, sofar, rest);
        
        while(cnt > 0){
            rest.insert(rest.begin(), x);
            cnt--;
        }
    }
};

File: leetcode/java/Subsets_2.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        if(nums == null || nums.length == 0) return Collections.emptyList();
        
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        sdh(result, new ArrayList<>(), nums, 0);
        return result;
    }
    
    private void sdh(List<List<Integer>> result, List<Integer> sofar, int[] nums, int s) {
        if(s == nums.length) {
            result.add(new ArrayList<>(sofar));
            return;
        }
        
        int e = s;
        while(e < nums.length-1) {
            if(nums[e] == nums[e+1]) e++;
            else break;
        }
        e++;
        
        sdh(result, sofar, nums, e);
        
        int i = s;
        int x = nums[s];
        while(i < e) {
            sofar.add(x);
            sdh(result, sofar, nums, e);
            i++;
        }
        
        while(i > s) {
            sofar.remove(sofar.size()-1);
            i--;
        }
    }
}
